### ISSUE TITLE
runscript not found when obtained from Docker

### ISSUE THREAD

**Issue by vsoch** [apptainer/singularity#422]
If I am importing an image from docker, the runscript file is generated as `/singularity` but the container does not find it. Eg.

      sudo singularity create mongo.img
      sudo singularity import mongo.img docker://mongo:latest

blabla bootstrap. Then:

      $ ./mongo.img 
      No Singularity runscript found, executing /bin/sh 
      Singularity.mongo.img> $ ls /singularity
      /singularity

If I bootstrap this same container but define it myself (via the bootstrap spec below) it does work. First, here is the spec:

      BootStrap: docker
      From: mongo:latest

      %runscript

      exec /entrypoint.sh "$@"

and then creating:

      sudo singularity create mongo.img
      sudo singularity bootstrap mongo.img Singularity

and then running:

      $ ./mongo.img --help
      Options:

      General options:
        -h [ --help ]                         show this usage information
        --version                             show version information
        -f [ --config ] arg                   configuration file specifying 

Probably is an issue with making the file executable, or something like that? We still need to finish up the bootstrap --> .env folder --> docker bit, so when we do that, I can look into this.

**Comment by eshelman** [comment]
I just hit this same bug. Indeed, it is the fact that the /singularity script is not executable. For now, you can fix images by running:

```singularity exec --writable mongo.img chmod 755 /singularity```

**Comment by vsoch** [comment]
yeah, hopefully we will get this fixed up soon! It will be wrapped in with the environment / bootstrap upgrades, which @bauerm97 worked a ton on, and @gmkurtzer is finishing up!

**Comment by vsoch** [comment]
Actually let me see what I can figure out now - this is a pretty important issue. I used to handle the permissions change in the docker portion and it was moved to postbootstrap to be consistent, so something must be not working there.

**Comment by vsoch** [comment]
I think I see the issue! Give me a few mins.

**Comment by neezbeez** [comment]
I'm seeing a similar issue with a container that was exported using docker export.  The Dockerfile from which that was generated has a CMD stanza, but this doesn't seem to get transferred over to the Singularity image.  Is there something that needs to be done on the export or import side to make this work?

**Comment by vsoch** [comment]
Did you try with the `development` branch? This will be better detailed with docs, but it works as follows:

 - if the user defines a runscript in the `%runscript` section, this always takes priority
 - if the Dockerfile has an `ENTRYPOINT`, this is checked next.
 - if the Dockerfile doesn't have `ENTRYPOINT`, we fall back to `CMD`
 - if you are using an image that has both, but want to use `CMD`, then you should specify  `IncludeCmd: yes` in the bootstrap header
 - if there is no Docker stuffs, and no `%runscript`, then you get `/bin/sh`

Does this answer your question? If you could try with development, and then if the issue persists, run the command with `--debug` I would be happy to help!

**Comment by neezbeez** [comment]
Still no dice with the dev branch.  
```
Singularity.ipm_phase0_2.img> cat /.exec 
#!/bin/bash
. /environment
exec "$@"
```
```
Singularity.ipm_phase0_2.img> cat /.run
#!/bin/bash
. /environment
if test -x /singularity; then
    exec /singularity "$@"
else
    echo "No Singularity runscript found, executing /bin/sh"
    exec /bin/sh "$@"
fi
```

There's no `ENTRYPOINT` in the Dockerfile, only a `CMD`.




[singularity_dev_debug.txt](https://github.com/singularityware/singularity/files/870242/singularity_dev_debug.txt)

**Comment by vsoch** [comment]
can you send the complete debug? I need to see the python bits too :) Thank you!

**Comment by neezbeez** [comment]
Just as a follow-up, I would be nice if this worked out of the box, but creating the `/singularity` file within the container from `singularity shell -w` and making it executable is a fine work-around for now.

**Comment by neezbeez** [comment]
That's all the debug that got printed.  Does it drop a log file elsewhere?

**Comment by vsoch** [comment]
could you provide the command you are using to generate this image? There should be python output, because the entire communication with anything and everything Docker is done with our python bits!

**Comment by neezbeez** [comment]
```
vanessa@myhost ~ $ sudo singularity create -s 8000 test.img
[sudo] password for vanessa: 
Creating a new image with a maximum size of 8000MiB...
Executing image create helper
Formatting image with ext3 file system
Done.
vanessa@myhost ~ $ sudo singularity --debug import test.img phase0.tar 
enabling debugging
ending argument loop
Exec'ing: /usr/local/libexec/singularity/cli/import.exec test.imgVERBOSE [U=0,P=18212]      message.c:52:message_init()                : Set messagelevel to: 5
VERBOSE [U=0,P=18212]      config_parser.c:113:singularity_config_parse(): Initialize configuration file: /usr/local/etc/singularity/singularity.conf
DEBUG   [U=0,P=18212]      config_parser.c:125:singularity_config_parse(): Starting parse of configuration file /usr/local/etc/singularity/singularity.conf
- snip -
```
The output from that point forward is identical more or less to what I posted earlier.

**Comment by vsoch** [comment]
Oh I see, you are trying to import a tar of a docker image? We don't have support for that. What you would want to do is one of the following:

     singularity create -s 8000 test.img
     singularity import test.img docker://phaser0:latest
     
or do the same with a bootstrap spec, `Singularity`

    Bootstrap: docker
    From: phaser0:latest
  
and then the command

      singularity create -s 8000 test.img
      sudo singularity bootstrap test.img Singularity  

For each of the above, the docker image is served on Docker hub. But if you are running a local docker registry, you can simply modify the uri to use it:

     singularity run docker://localhost:5000/phaser0

where 5000 is the port. Does that make sense? The reason you don't see the runscript / cmd is because those things are served as metadata from the docker hub API, and we add them to the container with the functions above. If you just dump the contents of the docker image, we don't do that step, which is why you don't even see python involved.

**Comment by vsoch** [comment]
vanessaaa!! Your prompt looks like mine!! lol

        vanessa@vanessa-ThinkPad-T460s:~$

**Comment by neezbeez** [comment]
LOL!! It's been making me laugh so much as I've been working through the documentation! So nice to meet another Vanessa who's an HPC nerd!

For the docker import, it works fine when pulling images from docker.io, but when using a local repo, it fails. This is still the dev branch build. I ran the import as root to rule out any weird sudo env issues.
[debug.txt](https://github.com/singularityware/singularity/files/870528/debug.txt)


```
[root@myhost ~]# export SINGULARITY_DOCKER_REGISTRY='--registry localhost:5000'
[root@myhost~]# singularity import test.img docker://nanocall:latest
Traceback (most recent call last):
  File "/usr/local/libexec/singularity/python/cli.py", line 290, in <module>
    main()
  File "/usr/local/libexec/singularity/python/cli.py", line 148, in main
    run(args)
  File "/usr/local/libexec/singularity/python/cli.py", line 229, in run
    auth=auth)
  File "/usr/local/libexec/singularity/python/docker/api.py", line 221, in get_manifest
    auth=auth)
  File "/usr/local/libexec/singularity/python/docker/api.py", line 72, in get_token
    response = api_get(base, default_header=False)
  File "/usr/local/libexec/singularity/python/utils.py", line 148, in api_get
    response = urlopen(request)
  File "/usr/lib64/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib64/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib64/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 1243, in https_open
    context=self._context)
  File "/usr/lib64/python2.7/urllib2.py", line 1200, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:661)>
ERROR: Aborting with RETVAL=1
```

**Comment by vsoch** [comment]
Ah we are getting closer! Another case of bad outdated documentation (looks sheepishly away, lol) Let's try something like this - does your image have a namespace/reponame?

```bash
singularity --debug import test.img docker://localhost:5000/nanocall/nanocall:latest
enabling debugging
ending argument loop
Exec'ing: /usr/local/libexec/singularity/cli/import.exec test.imgDEBUG:python:SINGULARITY_COMMAND_ASIS found as False
DEBUG:python:SINGULARITY_ROOTFS not defined (None)
DEBUG:python:SINGULARITY_METADATA_FOLDER found as None/.singularity
DEBUG:python:SINGULARITY_DISABLE_CACHE found as False
DEBUG:python:SINGULARITY_CACHEDIR found as /home/vanessa/.singularity
DEBUG:python:SINGULARITY_ENVBASE found as None/.singularity/env
DEBUG:python:SINGULARITY_LABELFILE not defined (None)
DEBUG:python:SINGULARITY_PULLFOLDER found as /home/vanessa/Desktop
DEBUG:python:SINGULARITY_CONTENTS found as /tmp/.singularity-layers.VOTCCwWI
DEBUG:python:SINGULARITY_CONTAINER found as docker://localhost:5000/nanocall/nanocall:latest
DEBUG:python:Found uri docker://
...
INFO:python:
*** STARTING DOCKER IMPORT PYTHON  ****
DEBUG:python:SINGULARITY_DOCKER_USERNAME not defined (None)
DEBUG:python:SINGULARITY_DOCKER_PASSWORD found
INFO:python:Root file system not found, writing layers to: /tmp/.singularity-layers.VOTCCwWI
DEBUG:python:Starting Docker ADD, only includes file and folder objects
```

the important lines you'd want to see are that it catches the correct registry, namespace and reponame

```bash
INFO:python:Docker image: localhost:5000/nanocall/nanocall:latest
INFO:python:Registry: localhost:5000
INFO:python:Namespace: nanocall
INFO:python:Repo Name: nanocall
INFO:python:Repo Tag: latest
INFO:python:Docker image path: localhost:5000/nanocall/nanocall:latest
INFO:python:Obtaining manifest: https://localhost:5000/v2/nanocall/nanocall/manifests/latest
```

**Comment by neezbeez** [comment]
Getting closer indeed! There's no repo namespace, unless "library" is created by default - I followed the steps here: https://github.com/singularityware/singularity/issues/429

Confirmed that the repo does work as expected:

```
[root@myhost ~]# docker pull localhost:5000/nanocall:latest
Trying to pull repository localhost:5000/nanocall ... 
sha256:58b2acffa5cb338de17bf2fbdcf35f9206539a6a6d1273c848c577b170a488c1: Pulling from localhost:5000/nanocall
Digest: sha256:58b2acffa5cb338de17bf2fbdcf35f9206539a6a6d1273c848c577b170a488c1
Status: Image is up to date for localhost:5000/nanocall:latest
```
The python bits of the import:
```
*** STARTING PYTHON CLIENT PORTION ****
INFO:python:Root file system defined as /var/lib/singularity/mnt/final
INFO:python:Including Docker command as Runscript? False
INFO:python:Docker image: nanocall:latest
INFO:python:Repo Name: nanocall
INFO:python:Repo Tag: latest
INFO:python:Namespace: library
INFO:python:Docker image path: library/nanocall:latest
INFO:python:Obtaining manifest: https://localhost:5000/v2/library/nanocall/manifests/latest
```
Env:
```
[root@myhost ~]# env | grep -i sing
SINGULARITY_DOCKER_NAMESPACE=--namespace 
SINGULARITY_DOCKER_REGISTRY=--registry localhost:5000
```

**Comment by vsoch** [comment]
Yes, library is the default namespace (Docker Hub standard), so in the case of a local registry you will have to set it by defining it with the image uri. Eg, `localhost:5000/namespace/repo_name:tag_name` Do you think there is a better way to go about this? We haven't had much of this use case, so if there is something more ideal than just defining it in the uri, let's chat about it!

**Comment by neezbeez** [comment]
That would be the ideal way to do it, imo.  That doesn't work unless you set the `SINGULARITY_DOCKER_REGISTRY` variable though as it will try to hit docker.io. That's counter-intuitive, and it should just parse the URL. 

I realized that the import is failing on a SSL error, which is the same thing I'm getting when I try to curl the catalog:

```
  File "/usr/lib64/python2.7/urllib2.py", line 1200, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:661)>
```
That's certainly outside of the scope of this conversation.  Thanks for all your help!

**Comment by vsoch** [comment]
haha ok, definitely ping again if you have more questions - just to be more clear (because the documentation is pretty much non existent) with the current version, the entirety of the registry, the namespace, repo, and tag is parsed via the uri. For example:

`docker://<registry>/<namespace>/<repo_name>:<repo_tag>` should work correctly for development, and if it doesn't we have a bug! We also just added support for a particular hash, like:

`docker://<registry>/<namespace>/<repo_name>@sha256:123456...`

and if the hash is found, it takes preference over the tag. I believe in Docker terms (as defined by their API docs) they call that final bit (tag or hash) a "digest".

The defaults that we set, if you don't define any of the above, are:

- registry: index.docker.io
- namespace: library
- repo_name: must be defined
- repo_tag: latest

and none for hash, since most user's use tags.

I promise to get this into updated docs soon! At least with the official release. :)
